### PR TITLE
Add ability to choose between URLSession and Alamofire

### PR DESCRIPTION
**Pull request by subdan**
A user can now choose between Alamofire and URLSession.

**CocoaPods**
Add two subspecs: Alamofire and URLSession. By default using Alamofire.

**SPM**
Before: one product — Apexy, uses Alamofire under the hood.
After: Three products: Apexy, Apexy_Alamofire, Apexy_RxSwift.

- Apexy uses URLSession under the hood.
- Apexy_Alamofire uses Alamofire under the hood.
- Apexy_RxSwift adds RxSwift support for URLSession or Alamofire.

I the PR I split `Client` to `AlamofireClient` (renamed) and `URLSessionClient` (new).

### CODE REVIEW

File: Sources/Apexy_URLSession/URLSessionClient.swift [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            completionHandler(.failure(error))
+            return Progress()
+        }
+        
+        let task = session.dataTask(with: request) { (data, response, error) in
+            let result = APIResult<T.Content>(catching: { () throws -> T.Content in
+                if let data = data {
+                    return try endpoint.content(from: response, with: data)
+                } else {
+                    throw error!
```
Comment (Alexander-Ignition) [inline Sources/Apexy_URLSession/URLSessionClient.swift]: there should be an error check not a data check

File: Sources/Apexy_URLSession/URLSessionClient.swift [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            request = try endpoint.makeRequest()
+            request = try requestAdapter.adapt(request).get()
+        } catch {
+            completionHandler(.failure(error))
+            return Progress()
+        }
+        
+        let task = session.dataTask(with: request) { (data, response, error) in
+            let result = APIResult<T.Content>(catching: { () throws -> T.Content in
+                if let data = data {
```
Comment (Alexander-Ignition) [inline Sources/Apexy_URLSession/URLSessionClient.swift]: `data ?? Data`

File: Package.swift [hunk 3]
```diff
@@ -6,19 +6,28 @@ import PackageDescription
...
     dependencies: [
-        .package(url: "https://github.com/Alamofire/Alamofire.git", .upToNextMajor(from: "5.2.0"))
+        .package(url: "https://github.com/Alamofire/Alamofire.git", .upToNextMajor(from: "5.2.0")),
+        .package(url: "https://github.com/ReactiveX/RxSwift.git", from: "5.0.0")
     ],
     targets: [
-        .target(name: "Apexy", dependencies: ["Alamofire"]),
+        .target(name: "Apexy_RxSwift", dependencies: ["Apexy", "RxSwift"]),
+        .target(name: "Apexy_Alamofire", dependencies: ["Apexy", "Alamofire"]),
+        .target(name: "Apexy_URLSession", dependencies: ["Apexy"]),
```
Comment (Alexander-Ignition) [inline Package.swift]: why aren't the words merged ?

File: Sources/Apexy_URLSession/URLSessionClient.swift [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    init(task: URLSessionUploadTask, inputStream: InputStream, completionHandler: @escaping (Result<Data, Error>) -> Void) {
+        self.task = task
+        self.inputStream = inputStream
+        self.completionHandler = completionHandler
+    }
+}
+
+private final class SessionDelegate: NSObject {
+    
+    private var tasks: [UploadTask] = []
```
Comment (Alexander-Ignition) [inline Sources/Apexy_URLSession/URLSessionClient.swift]: not thread safe

File: Sources/Apexy_URLSession/URLSessionClient.swift [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        var request: URLRequest
+        do {
+            request = try endpoint.makeRequest()
+            request = try requestAdapter.adapt(request).get()
+        } catch {
+            completionHandler(.failure(error))
+            return Progress()
+        }
+        
+        let task = session.dataTask(with: request) { (data, response, error) in
```
Comment (Alexander-Ignition) [inline Sources/Apexy_URLSession/URLSessionClient.swift]: URLSession closure based api and delegate api  are mutually exclusive.

You only need to use the dalaget api

File: Sources/Apexy_URLSession/URLSessionClient.swift [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        uploadTask.completionHandler(.failure(error))
+    }
+}
+
+// MARK: - URLSessionDataDelegate
+
+extension SessionDelegate: URLSessionDataDelegate {
+    
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
+        let uploadTask = tasks.first(where: { $0.task === dataTask })
```
Comment (Alexander-Ignition) [inline Sources/Apexy_URLSession/URLSessionClient.swift]: to do ``$0.task === dataTask``, there are `dataTask.taskIdentifier`

File: Sources/Apexy_URLSession/BaseRequestAdapter.swift [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import Foundation
+
+protocol RequestAdapter {
+    func adapt(_ urlRequest: URLRequest) -> Result<URLRequest, Error>
```
Comment (Alexander-Ignition) [inline Sources/Apexy_URLSession/BaseRequestAdapter.swift]: why not just `throws` ?

File: Sources/Apexy_URLSession/URLSessionClient.swift (hunk 1 above)
Comment (subdan) [reply Sources/Apexy_URLSession/URLSessionClient.swift]: What should I do if `error = nil` and `data = nil`? I get `Missing return in a closure` if I add an error check.
image

File: Sources/Apexy_URLSession/URLSessionClient.swift (hunk 2 above)
Comment (subdan) [reply Sources/Apexy_URLSession/URLSessionClient.swift:55]: Could you explain what do you mean?

File: Package.swift (hunk 3 above)
Comment (subdan) [reply Package.swift]: Initially I used symbol `+` but have some issues with it. And I replaced `+` with `_`. I think it is a good idea to merge the words.

**Comment by subdan** [comment]
I've removed `UploadEndpointBody.stream` implementation for `URLSessionClient` because it is very rarely used feature and it is hard to implement.

**Review by Alexander-Ignition — APPROVED** [review]

File: Sources/ApexyURLSession/URLSessionClient.swift [hunk 8]
```diff
@@ -0,0 +1,121 @@
+import Foundation
+import Apexy
+
+final public class URLSessionClient: Client {
```
Comment (vani2) [inline Sources/ApexyURLSession/URLSessionClient.swift]: Should we make it `open class` to let developers to add some functionality?

**Review by vani2 — APPROVED** [review]

File: Sources/ApexyURLSession/URLSessionClient.swift (hunk 8 above)
Comment (subdan) [reply Sources/ApexyURLSession/URLSessionClient.swift]: Yes, it’s a good idea to make it `open`.

File: Sources/ApexyURLSession/URLSessionClient.swift [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        do {
+            request = try endpoint.makeRequest()
+            request = try requestAdapter.adapt(request)
+        } catch {
+            completionHandler(.failure(error))
+            return Progress()
+        }
+        
+        let task = session.dataTask(with: request) { (data, response, error) in
+            let result = APIResult<T.Content>(catching: { () throws -> T.Content in
```
Comment (modestman) [inline Sources/ApexyURLSession/URLSessionClient.swift:52]: Please add a validation check `endpoint.validate`

**Review by modestman — CHANGES_REQUESTED** [review]

File: Sources/ApexyURLSession/URLSessionClient.swift (hunk 9 above)
Comment (subdan) [reply Sources/ApexyURLSession/URLSessionClient.swift:52]: Added

**Review by modestman — APPROVED** [review]

**Pull request merged**